tools/shared.py

```diff
@@ -1981,7 +1981,7 @@ def is_wasm_only():
     # fastcomp can emit wasm-only code.
     # also disable this mode if it depends on special optimizations that are not yet
     # compatible with it.
-    return 'asmjs' not in Settings.BINARYEN_METHOD and 'interpret-asm2wasm' not in Settings.BINARYEN_METHOD and not Settings.RUNNING_JS_OPTS and not Settings.EMULATED_FUNCTION_POINTERS and not Settings.EMULATE_FUNCTION_POINTER_CASTS
+    return 'asmjs' not in Settings.BINARYEN_METHOD and 'interpret-asm2wasm' not in Settings.BINARYEN_METHOD and not Settings.RUNNING_JS_OPTS and ((not Settings.EMULATED_FUNCTION_POINTERS and not Settings.EMULATE_FUNCTION_POINTER_CASTS) or not Settings.LEGALIZE_JS_FFI)
```

Comment:
If everything is grouped, then the is_wasm_only assert in emcc.py doesn't actually serve any purpose (basically a no-op). The EMULATED_FUNCTION_POINTERS and EMULATE_FUNCTION_POINTER_CASTS options don't seem to break the LEGALIZE_JS_FFI=0 mode so that's why I grouped those with it. In other words, with the narrower grouping, the assert will fire for non-wasm BINARYEN_METHOD or RUNNING_JS_OPTS when used with LEGALIZE_JS_FFI=0 and the EMULATE* options can be any setting.

If you want, I'm okay with just grouping it around the whole option and dropping the assert. The only downside I can see is that failures might be a bit less clear (and happen later) when using LEGALIZE_JS_FFI=0.

I'm new to the emscripten code base so some of this is trial and error until I get something that seems to work right (i.e. wasm/wast output that I'm expecting) so I apologize if it feels like I'm stabbing about a bit randomly (because I probably am).